cityofaustin/atd-moped
# Upgrade to react-scripts v5

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/9561

## Steps I took (for reference):
1. `$ nvm use` 😉 
2. delete `/node_modules` and `package-lock.json`
3. edit `package.json` to remove `react-map-gl-draw`
4. edit `package.json` to remove all `devDependencies`
6. manually set `react-scripts` in `package.json` to `^5.0.0`
7. re-install node modules and rebuild `package-lock.json`: `$ npm install`
8. install `react-map-gl-draw` again, forcibly: `$ npm install react-map-gl-draw@^0.22.3 --force`
9. manually bump react/react-dom `peerDependencies` of `react-map-gl-draw` to `^17.x` in `package-lock.json` 
10. install babel import plugin: `$ npm install --save-dev babel-plugin-import`
11. remove `eslintConfig` in `package.json` and update `.eslintrc.json`
12. code changes to get file upload and components map to compile
13. code changes to stop exporting anonymous objects
14. `$ npm run test` ⚙️   
13. `$ npm run start` ✨ 

## About our ESLint config

It's been hard for me to untangle the effects of this upgrade on linting. We're migrating from ESLint v6 to v8, as well as upgrading a number of related parsers and configs.

Also, as far as I can tell, the `react-app`  ESLint config was previously being ignored, because it was defined in `package.json`, and it was being overwritten by our `.eslintrc` file ([docs](https://eslint.org/docs/latest/user-guide/configuring/configuration-files)).

One issue I immediately encountered were errors with anonymous default exports. That rule has been in place [for years](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-anonymous-default-export.md), but was maybe only enabled by `react-app`? 🤔. Anyway, I addressed those errors in this PR.

The `react-app`  ESLint config is Create React App's base lint configuration, and they recommend that we do not ignore it. It's now enabled. Besides a few anonymous default export problems, the app is (surprisingly?) compiling without errors. 

Additionally, we had a number of ESLint plugins through [`eslint-config-airbnb`](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb). However, none of these configs were enabled in our `.eslintrc` file. I removed them:

```
 'eslint-plugin-import': '^2.25.3',
  'eslint-plugin-jsx-a11y': '^6.5.1',
  'eslint-plugin-react': '^7.28.0',
  'eslint-plugin-react-hooks': '^4.3.0'
```

And, lastly, i removed the `prettier` linting configs as well. I don't fully understand, but the latest version threw a lot of errors and clearly had different opinions than our previous setup. 

In summary, we need to revisit all of our linting soon.

## Other notes

- `npm install` will fail without our package-lock.json. we are stuck dealing with force installing `react-map-gl-draw` for the near future.
- we're getting an ugly compilation warning about `DoubleScrollbar.js`:

```shell
WARNING in ./node_modules/react-double-scrollbar/dist/DoubleScrollbar.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/atd-moped/moped-editor/node_modules/react-double-scrollbar/dist/DoubleScrollbar.js.map' file: Error: ENOENT: no such file or directory, open '/atd-moped/moped-editor/node_modules/react-double-scrollbar/dist/DoubleScrollbar.js.map'
```

Seems harmless, and it's possible to suppress the warning. [Discussed here](https://github.com/material-table-core/core/issues/467).

## Testing
**URL to test:** [deploy preview](https://deploy-preview-695--atd-moped-main.netlify.app/moped/projects) and local

### If testing locally
1. Remember to switch Node environments with`$ nvm use` 
2. You will probably need to delete `node_modules/` and then `$ npm install`, but who knows
3. Follow usual steps to start Moped editor + database.
4. Please make some code changes and test out the live-refresh.
6. try out the Jest tests, too: `$ npm run test`


### Things to test
1. Make sure File uploads work correctly
2. Make sure static images load correctly:
- Component map fallback image (when no components have been mapped)
- The Streets / aerials buttons on the component map

---
#### Ship list
